### PR TITLE
Add the webhook

### LINKED ISSUE #38: Update Webhook events documentation
Updating webhook events documentation to take into account the new events sent from Pre-Screen. 

The webhooks are the ones from the following issues: 

- https://github.com/CodeSignal/codesignal/pull/28841
- https://github.com/CodeSignal/codesignal/issues/28861

- `preScreenResultVerificationPending`.
- `preScreenExpired`.
- `preScreenCandidateDeclined`.
- `preScreenResultNotVerified`.
- `preScreenResultVerified`.

### PR DESCRIPTION
**Pull request by lizhoucheng**
Closes #38 

Added the following webhooks:

- companyTestSessionCreated
- preScreenResultVerificationPending.
- preScreenExpired.
- preScreenCandidateDeclined.
- preScreenResultNotVerified.
- preScreenResultVerified.

### CODE REVIEW

File: docs/webhooks/index.html [hunk 1]
```diff
@@ -253,6 +254,36 @@ <h2>Events for CodeSignal Test</h2>
  <div class="subsection">
 <h3 id="testworkflow">Test webhooks process flow</h3>
 <img alt="Diagram of webhooks process flow for CodeSignal Test" width="100%" src="https://codesignal.com/wp-content/uploads/2021/03/webhooksproces_test.png" />
+</div>
+<div class="subsection">
+    <h3 id="companyTestSessionCreated">companyTestSessionCreated</h3>
+    <p>
+        Fired when a client creates a test session.
```
Comment (Albinutte) [inline docs/webhooks/index.html:261]: I would rephrase this to say "Fired when a test session is created" -- it can be created in a numerous of ways, including automatic creation through the ATS or when a candidate follows a public URL.

File: docs/webhooks/index.html [hunk 2]
```diff
@@ -253,6 +254,36 @@ <h2>Events for CodeSignal Test</h2>
...
+event: 'companyTestSessionCreated',
+triggeredOn: 1553720789347,
+payload: {
+testSessionId: 'lehu382hdleh29',
+testId: 'dhey29hcl28hl28',
+testTitle: 'Software Engineer intern',
+candidateEmail: 'john.doe@gmail.com',
+candidateName: 'John Doe'
+}
+};</code></pre>
```
Comment (Albinutte) [inline docs/webhooks/index.html:286]: Could you make sure the JSON has proper indentations? This doesn't look great atm:
image

**Review by Albinutte — COMMENTED** [review]
Thank you for updating the docs! Left a couple of comments, please check them out.

Regarding the graph, I see that the last one was added by @malpine2. @malpine2, perhaps you could help us with updating it?

**Comment by miamagana** [comment]
Hello @lizhoucheng, I have been working on new Webhooks and have created the following diagram, could we use it to update the docs once? Let me know if you agree.

company test flow-2 drawio-2

**Comment by malpine2** [comment]
> Thank you for updating the docs! Left a couple of comments, please check them out.
> 
> Regarding the graph, I see that the last one was added by @malpine2. @malpine2, perhaps you could help us with updating it?

Looks like @miamagana has this covered? It's been a while since I've worked on these docs!

**Comment by miamagana** [comment]
> Looks like @miamagana has this covered? It's been a while since I've worked on these docs!

I can update the graphs taking into account @lizhoucheng changes. However I would advise to store them on this repo for further modifications.

**Comment by lizhoucheng** [comment]
@miamagana Feel free to combine your changes to this PR! And thank you for containing the `companyTestSessionCreated` event into the graph if not too much work!

**Comment by miamagana** [comment]
> @miamagana Feel free to combine your changes to this PR! And thank you for containing the `companyTestSessionCreated` event into the graph if not too much work!

Of course, will do it!

**Comment by miamagana** [comment]
Hello @Albinutte @lizhoucheng I have updated this PR with my changes and taking into account the new `preScreenResultVerified` event to be created in #28861.

Mostly, I have separated the flows by the actor (candidate, codesignal auditor and company client) and added the new events.

File: docs/webhooks/index.html [hunk 3]
```diff
@@ -328,7 +449,127 @@ <h3 id="companyTestSessionFinished">companyTestSessionFinished</h3>
...
-		<h2>Events for CodeSignal Certify</h2>    
+<div class="subsection">
+   <h3 id="auditorworkflow">CodeSignal Auditor workflow</h3><br>
+       <img alt="Diagram of webhooks triggered on CodeSignal auditor flow" width="100%" src="../images/auditor-events.png" />
+</div>
+
+                <div class="subsection">
+                    <h4 id="preScreenResultNotVerified">preScreenResultNotVerified</h4>
+                    <p>
+                        Fired when an auditor rejects a proctored test session.
```
Comment (Albinutte) [inline docs/webhooks/index.html:460]:
```suggestion
                        Fired when a CodeSignal auditor does not verify a proctored test session.
```

Not sure about this one, sounds a bit funny 🤔

**Review by Albinutte — COMMENTED** [review]
Awesome, thank you for doing this! It looks like I can't comment directly on the images, so I'll leave my comments here instead:
* `auditor-events`:
    * Perhaps it would make sense to keep `preScreenResultNotVerified` webhook in `auditor-events` green, similar to all other webhooks. I would also suggest replacing `rejected` with `not verified`, since we're trying to move away from "reject" wording in the main app and when interacting with clients.
* `candidate-events`
    * I would perhaps suggest removing explanation in the rectangular blocks -- I'm not sure if the implementation details are important to the customers.
    * Should there be a "Yes" arrow from "starts screen" to "companyTestSessionStarted"?
    * I would suggest replacing "screen" with "test session", since this is the term we're using in the main app.
* `client-events`
   * A test session can technically be created in many different ways -- ATS/GraphQL API/UI/when a candidate follows a public link. Showing just "company client" on the chart seems a bit misleading.

File: docs/diagrams/webhooks-flow [hunk 4]
```diff
@@ -0,0 +1 @@
...EqyqHkzOpHIw/gt0bgp0MYocbO6OfaGfVa9IzRzvprEdUNh/l61scCEwGVcDK567Y+ZHNgfKkVkTooWWzkhvsLZ2VTjH4QHqCP0+O9GfT7CYgOjU3gHXznjkgSoVUgOKTaO3FW85xPaI0xSTOofO/qGEY7HqvYUz4XnDWfcEiYXcYn1Y/+LTcjgsWyeBWElmRv9iJPJstcIzTFQengTDr1xK2MecT93dcp0FKKcURF4Q4wjpvrTrZ+XOg7dpq8HMacmATkzk+Ww60P7B5CHKHsdl6LXgcNstxV9h8bzSp9O52eFJouIr6h0+VmnpX0C97MOzm73jAj7sB/LgGeNuyuwW8/mL4PiH8dSF3hn77SQfsYVYqudRtSKX1eqa9VKpTdXpncVtWL++utRg7WmXYf2uhd3MlJf2A0iykf6MEL1iW4e+y7xGrEOPlO9R9va1zGsiTbjUP9cUGZZ2ONmucNPj4R2VZY58P09QdsTJ1mrr0rwOJpstiBupU9h+LkXxnV2kS499o7neCwxF9YH4UCwC8O9ydlrtqHhzYoGgeZ+VOP8nFz2SeIy1/xMK3+X36bychHFo7Kf21KR7f7IkKw4++JQU9nOjb6wqk01d8nhA6j13PAIe1wgjq6w5mJGCN7pJZgMmvsgDzZXh1s77rdXquoQOl04V6cDwcnAyuCLGPRVJH9BaExqVhjR0wkf8rzjTpmImGOttVNBN7KtthUF7S49uhvExHXQItlxgPl/KSEeiD3zoB8VZHbYddK6xcc3I5qYaATgFv6S3tA3pmiBk04KXDt8Rx11qKUwEvLwPVM87YJeqLFAVlmTM/XE0lH9Q3hVlykXKBp9v3Ck94JJ/2Pts4W/XtScNnJMBuUHd9hXJr6pZxdr+67wjX+AK3rfGF6yPfb9L43LUtu6V+wKnRSm5T49zUELSh8dXDdD6b3l2svm228E8jtqbWRb9fSqrVAaz9vLmJPq2AqOX2gxVFozobvqNsPM46T/lQcy8AJqx+sAAYO9bYsF23lnXaqx8aJ9nQ4SfIo+VHlLT4iDIt3gT1V1IcXSmoPrEfT69VFMCAtRbCHG5F8ZqPK8RN/lqPCWchRrFOBHdN2jRYVEPz0aQNAG53rM3+aHHgRWoBrZ7KkJP0bdc/9m7SN2h6df0LWZENd3+boEDE7i88wJv/AA==</diagram></mxfile>
```
[the remark this answers is not in the document]
Comment (miamagana) [reply docs/diagrams/webhooks-flow:1]: This is the diagram in xml exported from draw.io to be able to edit it in the future, I can exclude it, but I feel like it would make much easier to modify it.

File: docs/webhooks/index.html [hunk 5]
```diff
@@ -285,9 +367,48 @@ <h3 id="companyTestSessionStarted">companyTestSessionStarted</h3>
...
+    score: 900,
+    maxScore: 1000,
+    codingScore: 820
+  }
+};</code></pre>
+                </div>
+                <div class="subsection">
+                    <h4 id="companyTestSessionFinished">companyTestSessionFinished</h4>
+                    <p>
+                        Fired when a candidate finishes a non proctored test session (including if the session gets force-finished because the time runs out).
```
[the remark this answers is not in the document]
Comment (miamagana) [reply docs/webhooks/index.html:411]: Yeah, you're right.

File: docs/webhooks/index.html (hunk 3 above)
Comment (miamagana) [reply docs/webhooks/index.html:460]: How can we say reject without reject? Declines?

Comment (Albinutte) [reply docs/webhooks/index.html:460]: "Marks session as not verified"? Englishsky is hard 😅  Perhaps @menewman has an idea?

**Comment by lizhoucheng** [comment]
Thank you for the changes! Please update the issue description before merging as well :)

File: docs/webhooks/index.html (hunk 3 above)
Comment (menewman) [reply docs/webhooks/index.html:460]: Yeah, it's tricky because "does not verify" can sound passive (like they took _no_ action or didn't review it at all), and we want to convey that they took the _proactive_ step of flagging a session for suspicious behavior.

I agree it helps to rephrase this just as you suggested, with "marks ... as not verified" conveying that it's an _active_ thing, not passive.

e.g., "Fired when a session is marked as not verified by CodeSignal."

If we want to elaborate a little, we could append:
"This means that either the session has been flagged for suspicious behavior, or a technical issue has been identified during the session (in which case the candidate will be automatically granted a retake)."

Note that I took out the word "auditor" from my version. That's intentional, because it might be nice to obscure the actual verification mechanism in cases where a human may or may not have actually reviewed the session.

File: docs/webhooks/index.html [hunk 6]
```diff
@@ -251,11 +269,75 @@ <h3 id="payloads">Webhook payloads</h3>
...
+                    <pre><code>{
+  event: 'preScreenExpired',
+  triggeredOn: 1553720789347,
+  payload: {
+    testSessionId: 'lehu382hdleh29',
+    testId: 'dhey29hcl28hl28',
+    testTitle: 'Software Engineer intern',
+    candidateEmail: 'john.doe@gmail.com',
+    candidateName: 'John Doe',
+    expiredReason: 'testNotTaken'
```
Comment (menewman) [inline docs/webhooks/index.html:335]: [related to my comment above] This feels like either it should be human-readable _or_ we should provide an enumerated list of possible "terse" versions. Right now it's a slightly cryptic code, which will be a bit hard for consumers to interpret without context.

File: docs/webhooks/index.html [hunk 7]
```diff
@@ -328,7 +449,127 @@ <h3 id="companyTestSessionFinished">companyTestSessionFinished</h3>
...
+                    <pre><code>{
+  event: 'preScreenResultNotVerified',
+  triggeredOn: 1553720789347,
+  payload: {
+    testSessionId: 'lehu382hdleh29',
+    testId: 'dhey29hcl28hl28',
+    testTitle: 'Software Engineer intern',
+    candidateEmail: 'john.doe@gmail.com',
+    candidateName: 'John Doe',
+    rejectedReasons: ['Two humans in the room', 'Screen not visible]
```
Comment (menewman) [inline docs/webhooks/index.html:485]: I think it would be ideal to use examples of our "real" rejection reasons here if possible. I think "Screen not visible" is realistic but I don't believe "Two humans in the room" is exactly the phrasing we use on prod?

File: docs/webhooks/index.html [hunk 8]
```diff
@@ -328,7 +449,127 @@ <h3 id="companyTestSessionFinished">companyTestSessionFinished</h3>
...
+    }
+};</code></pre>
+                    </div>
+
+                <div class="subsection">
+                    <h4 id="preScreenResultVerified">preScreenResultVerified</h4>
+                    <p>
+                        Fired when a CodeSignal auditor verifies a proctored test session.
```
Comment (menewman) [inline docs/webhooks/index.html:493]: Echoing my comment on the other thread, maybe this could be:
> Fired when a proctored session is verified by CodeSignal.

File: docs/webhooks/index.html [hunk 9]
```diff
@@ -328,7 +449,127 @@ <h3 id="companyTestSessionFinished">companyTestSessionFinished</h3>
...
+     testId: 'dhey29hcl28hl28',
+     testTitle: 'Software Engineer intern',
+     candidateEmail: 'john.doe@gmail.com',
+     candidateName: 'John Doe'
+   }
+ };</code></pre>
+</div>
+
+
+		<h2>Events for CodeSignal Certify</h2>
```
Comment (menewman) [inline docs/webhooks/index.html:572]: Maybe out-of-scope, but might be nice to update "CodeSignal Test" to "CodeSignal Pre-Screen" and "CodeSignal Certify" to "CodeSignal Pre-Screen Certify" where it's possible to do so within this document?

File: docs/diagrams/webhooks-flow.drawio [hunk 10]
```diff
@@ -0,0 +1 @@
...qyrfZY1nlMLiQyuXNTliofrGREoodh7ugLT7y/X0fQ5fc/ZZnP54/Dz///BM8Sz7S/QXMqLVzbFX/AgPgje8OgGMXfwMlqB1wseXGaNjgVLLSCM88mT+qFt3vsod4JrTAQKWa8Dyq2QC39DRzPTIHoZwbPKIs2+TH3WIF27psE7TA6AgdqlheknzXg3nBieuT7bS5HOkkuaB5/DweC1Oc4GxuYHlXiWT2MSQDnlMJus4Dmb7uTMhiGSbrP1hmI6BsA23XSdUddVW3ods+VXf/Hzj4FtiznVNhr3ZabiaJnqs6kOc2W4Wwr60AKb27ngoklLvQNRUhLw+ots4M2606djsPlwF/6TgeeJX0LghugvK/ZqNa33x9SAnfbeMWbzdv8EBFU56eN/iG9dK/VDVRNmwcpCj3+4hm9zmK5pp+B8qJOi3clnC/oCyP6V8oxVO2lPCt4AeURDiZXaHuq/zQMyTJzWoe2vs1f7da4vQqfRxaTtf03OpW/fn9U29fcuGpO32wW1vxUC82jEnEeOuMWYDJh3mEeb9K1axX3dDiWy13tMBBd+Kka9mqDJDjkdRQtgqhdWM5A0bh5F/FmYIqR7v0FrqevBa0oc2C4NEMVnp8PRms7amdhLIXr7sZrESgX+nqGRHv2M1EvBtU1gub5bVlCGg04B290fQToUUsdT3spdvXVLiSnlJT2Fd2Ri+PAlBv4hElik2mJyxa5H9tlyy2tQfZAOW33QEFT8yfrzKZs13VtwND1zVwDequdvDVpm6JtvvVzWCIaXyKr1PjXqUX0jelz6a0oA6Nz7+MnsajT/3599Ua/m4lzsjpt7uX3xmOYNwlNnzPyKjChirfbsV3AOClbweAgecMLNf3K0vKHoqgXRcMtO8vVZyt0FQdZOOQ/g6WhccpCq9057r69TDgG1DWVIivI+YPOXZJ4UPu55tvhVmZ6HoSjR+To4q3TqUGBkKmaXb/WgXcihoHBjWaKo91VFSMauxQ4bFLbVBGXZ3aDlBUzdvqgtI3qFrdo6+z5WlP622peVMJL0ELOlHcOORQJVPJTZXzc5izgAoO1FqdY6B4l9p2NWqqa9+eehvp6xJy7Z+31yusG9lP87qQGqZpuC5NEFCxlyRuN/ml97mVXyQ4Mh8GR+YD59B8dlA8cY14wIa7X2gopu9+6ALe/Qc=</diagram></mxfile>
```
Comment (menewman) [inline docs/diagrams/webhooks-flow.drawio:1]: Is this file added intentionally?

Comment (menewman) [reply docs/diagrams/webhooks-flow.drawio:1]: Same question about the pyc file, although I can't leave a comment on it directly since it's a binary file 😄

**Review by menewman — COMMENTED** [review]
Since I got a ping on one of the threads here, I went ahead and did a review on the whole PR 😄 

I don't necessarily need to re-review this, but please consider!

File: docs/diagrams/webhooks-flow.drawio (hunk 10 above)
Comment (miamagana) [reply docs/diagrams/webhooks-flow.drawio:1]: This file is the diagram, so whenever another dev has to edit anything in the future they won't need to create it from scratch like I did. Should I remove it?

File: docs/diagrams/webhooks-flow (hunk 4 above)
[the remark this answers is not in the document]
Comment (Albinutte) [reply docs/diagrams/webhooks-flow]: Oh in this case I agree, it would be nice to keep diagram! Could you also update the README to explain how the diagram can be updated?

**Review by Albinutte — APPROVED** [review]
Super minor, but could you replace "screen" with "test session" here?
image

I also generally agree with Michael's comments. They seem minor enough, but I'd like to take a look at the changes should you chose to update the docs with enumeration reasons for `expiredReason` just to see the format.

File: docs/webhooks/index.html (hunk 3 above)
Comment (miamagana) [reply docs/webhooks/index.html:460]: I like that.

File: docs/webhooks/index.html [hunk 11]
```diff
@@ -251,11 +269,75 @@ <h3 id="payloads">Webhook payloads</h3>
...
+                    <pre><code>{
+  event: 'preScreenExpired',
+  triggeredOn: number,
+  payload: {
+    testSessionId: string,
+    testId: string,
+    testTitle: string,
+    candidateEmail: string,
+    candidateName: string,
+    expiredReason: string,
```
[the remark this answers is not in the document]
Comment (miamagana) [reply docs/webhooks/index.html:321]: Yeah, just two. Let me add them.

**Comment by miamagana** [comment]
Hello @menewman I think I addressed everything, please let me know if I can do something else prior to merging this!

File: docs/diagrams/webhooks-flow.drawio (hunk 10 above)
Comment (menewman) [reply docs/diagrams/webhooks-flow.drawio]: Ah, sorry that I repeated Albina's comment from her review. I agree with her that it might be helpful to mention this in the README to clarify what it's for. But yeah, that justification is fine with me 👍

Comment (menewman) [reply docs/diagrams/webhooks-flow.drawio]: ...which, you did! So, we're all good 😄

File: docs/webhooks/index.html [hunk 12]
```diff
@@ -318,10 +318,16 @@ <h4 id="preScreenExpired">preScreenExpired</h4>
     testTitle: string,
     candidateEmail: string,
     candidateName: string,
-    expiredReason: string,
+    expiredReason: 'notCertified' | 'testNotTaken', *
```
Comment (menewman) [inline docs/webhooks/index.html:321]: Ah, I should have realized this sooner. Should we be sending "not verified" instead of "not certified" as the expired reason here?

**Review by menewman — APPROVED** [review]
👌 Changes look great, thank you! I just have one more comment/question

File: docs/webhooks/index.html (hunk 12 above)
Comment (miamagana) [reply docs/webhooks/index.html:321]: I think we should change it on the new web hook. Changing it on the pre existing ones might break customer code. 

So if we change it on the new pre-Screen expired webhook we would be moving away from Certify wording once we sunset Certify.

Comment (miamagana) [reply docs/webhooks/index.html:321]: Actually, we could just remove the 'notCertified' option, because that's no longer the case in Pre-Screen Test. The graph is correct and the only way a Pre-Screen Test can expire now is if the candidate doesn't take it.

**Comment by miamagana** [comment]
> @miamagana - I know it already merged, but one final thought -- if there's only one reason for expiration, perhaps we should just get rid of the expired reason field entirely? It doesn't seem to add much value now.

It fully makes sense, I can change it for sure.

**Comment by menewman** [comment]
Awesome, thank you!
